faster_coco_eval/core/boundary_utils.py

```diff
@@ -0,0 +1,114 @@
+"""
+Copyright (c) 2021, Bowen Cheng
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met: 
+
+1. Redistributions of source code must retain the above copyright notice, this
+   list of conditions and the following disclaimer. 
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution. 
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+The views and conclusions contained in the software and documentation are those
+of the authors and should not be interpreted as representing official policies,
+either expressed or implied, of the FreeBSD Project.
+"""
+import multiprocessing
+import time
+
+import cv2
+import numpy as np
+
+import pycocotools.mask as mask_utils
+
+
+# General util function to get the boundary of a binary mask.
+def mask_to_boundary(mask, dilation_ratio=0.02):
+    """
+    Convert binary mask to boundary mask.
+    :param mask (numpy array, uint8): binary mask
+    :param dilation_ratio (float): ratio to calculate dilation =
+                                   dilation_ratio * image_diagonal
+    :return: boundary mask (numpy array)
+    """
+    h, w = mask.shape
+    img_diag = np.sqrt(h ** 2 + w ** 2)
+    dilation = int(round(dilation_ratio * img_diag))
+    if dilation < 1:
+        dilation = 1
+    # Pad image so mask truncated by the image border is 
+    # also considered as boundary.
+    new_mask = cv2.copyMakeBorder(
```

Comment:
Can we stop using opencv? For example, implement this function yourself.

I would like to decouple the core library from everything related to large libraries.

Reply:
new_mask= np.pad(array=mask, pad_width=[[1, 1], [1, 1]], mode='constant', constant_values=0)

like this?

Reply:
Let me check what we can do here. I agree that a pure numpy implementation would be really cool. The padding is easy, the 'hard' part is the erode call, as you pointed out below. I have a little test notebook where I can tinker around.

Reply:
So, I tested an equivalent implementation using only numpy. While it works, it's very slow in comparison. The cv2 implementation runs at `~1000 it/s`, [kornia (pytorch)](https://github.com/kornia/kornia/blob/aec0dd863fc1e80ba2ded4fdbd2399e1f4631d46/kornia/morphology/morphology.py#L108) at `~50 it/s` and a pure numpy version of kornias erosion (using numpy as_strided) at around `~1.5 it/s`. Most likely, the runtime can be improved, but at the moment, I don't have the time to optimize this or to implement a more efficient algorithm.

I therefore propose to move boundary_utils to extra or to a different place to keep the core library free of large dependencies. IMHO, it makes no sense to pair a faster boundary eval with a slower boundary conversion. What do you think?

Reply:
Hello! I have an idea for this, let's create extras_require for Boundary. The installation will be like pip3 install faster-coco-eval[boundary]. Then we can implement dependencies, for example opencv.

But in this case, you need to implement a mechanism whereby if a person runs iouType=='boundary', you need to check that the opencv library is installed.

I can suggest a similar implementation
https://github.com/MiXaiLL76/faster_coco_eval/blob/main/faster_coco_eval/extra/utils.py#L1

And with opencv_available==False, it will throw an error with the correct library installation path!

Reply:
Hey, sry for the late reply. I'll have some time later this week so can finish this. I think this would be a good solution.

Reply:
Not sure if still relevant but I moved boundary_utils to extra